Adding Umbraco to "Use a CMS with Astro" guide (#7879)
withastro/docs: src/content/docs/en/guides/cms/umbraco.mdx

---
const res = await fetch('http://localhost/umbraco/delivery/api/v2/content?filter=contentType:article');
const articles = await res.json();
---
<h1>Astro + Umbraco 🚀</h1>
{
  articles.items.map((article) => (
      <h1>{article.name}</h1>
      <p>{article.properties.articleDate}</p>
      <div set:html={article.properties.content?.markup}></div>
  ))
}
```

<ReadMore>Read more about [data fetching in Astro](https://docs.astro.build/en/guides/data-fetching/).</ReadMore>

## Building a blog with Umbraco and Astro

### Prerequisites

- **An Astro project** - If you don’t have an Astro project yet, our [Installation guide](/en/install/auto/) will get you up and running in no time.

- **An Umbraco project (v12+)** with the Content Delivery API enabled - Follow this [Installation guide](https://docs.umbraco.com/umbraco-cms/fundamentals/setup/install/) to create a new project.

### Creating blog posts in Umbraco

From the [Umbraco backoffice](https://docs.umbraco.com/umbraco-cms/fundamentals/backoffice), create a Document Type for a simple blog article called 'Article'.

<Steps>
1. Configure your 'Article' Document Type with the following properties:

> - Title (DataType: Textstring)
> - Article Date (DataType: Date Picker)
> - Content (DataType: Richtext Editor)

Reviewer: ```suggestion
		- Title (DataType: Textstring)
		- Article Date (DataType: Date Picker)
		- Content (DataType: Richtext Editor)
```

---
title: Umbraco & Astro
description: Add content to your Astro project using Umbraco as a CMS
type: cms
stub: false
service: Umbraco
i18nReady: true
---

> import { FileTree } from '@astrojs/starlight/components';

Reviewer: ```suggestion
import { FileTree, Steps } from '@astrojs/starlight/components';
import ReadMore from '~/components/ReadMore.astro'; 
```

For more details on Data Fetching with Astro please read the [official documentation](https://docs.astro.build/en/guides/data-fetching/).

## Building a blog with Umbraco and Astro

### Prerequisites:

- Astro project, newly created with `npm create astro@latest` command with the following settings:
	- Where should we create your new project? `./astro.frontend`
	- How would you like to start your new project? `Empty`
 	- Do you plan to write TypeScript? `Yes`
  	- How strict should TypeScript be? `Strict (recommended)`
  	- Install dependencies? `Yes`
  	- Initialize a new git repository? `No`

- Umbraco project (v12+) - please follow this [Installation guide](https://docs.umbraco.com/umbraco-cms/fundamentals/setup/install/).
	- Enable the Content Delivery API. Instructions have been provided in the "Setting up the Content Delivery API" section above.

Once Astro has been setup and Umbraco installed, you are now able to create a blog that uses Umbraco as the CMS.

### Creating the blog posts in Umbraco

> From the Umbraco backoffice, create a Document Type for a simple blog article called 'Article'.

Author: Yeah "backoffice" is just the Umbraco name for "admin area" or similar!

To follow along below, your 'Article' Document Type should have the following properties:

- Title (DataType: Textstring)
- Article Date (DataType: Date Picker)
- Content (DataType: Richtext Editor)

For this simple example, on the 'Article' document type, set the permission "Allow as root" to `true`.

> Please create and publish some test articles.

Author: So this should probably be:

3. From the 'Content' section in the Umbraco backoffice, [create and publish your first blog post](https://docs.umbraco.com/umbraco-cms/fundamentals/data/defining-content). Repeat the process as many times as you like.

> Please create and publish some test articles.

Reviewer: Great thanks! I've updated my suggestion! (It would have been fine if you just edited your own text directly, too. Either works!)

Once these articles are published you should be able to retrieve the article details for all published articles as a JSON object via a GET request to the following URL:

`http://localhost:44335/umbraco/delivery/api/v2/content?filter=contentType:article`

> Please replace the port number with your local development environment's HTTP endpoint port number.

Author: Fair! Yes, this step is just checking that the content (in JSON format) can be retrieved with a GET request, which is what Astro will eventually be doing for us.

You don't *need* to carry out this step. Happy to go with your guidance on whether to go with a) omit or b) reframe

> Please replace the port number with your local development environment's HTTP endpoint port number.

Reviewer: OK then! I'm going to suggest we leave this out entirely then, and just close off our steps!

For information on creating Document Types, see [this guide](https://www.youtube.com/watch?v=otRuIkN80qM).

### Displaying a list of blog posts in Astro

Within our Astro project, this is the folder structure we will be using to create our blog:

<FileTree>
- src/
  - layouts/
    - Layout.astro
  - pages/
    - index.astro
    - [...slug].astro
</FileTree>

If using the `Empty` project, please create the `layouts/` folder, and `Layout.astro` and `[...slug].astro` files.

For more information on Layouts see the [Astro documentation](https://docs.astro.build/en/basics/layouts/).

Add the following markup to the `Layout.astro` file:

```astro title="src/layouts/Layout.astro"
---
---
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>My Blog with Astro and Umbraco</title>
</head>
<body>
    <main>
        <slot />
    </main>
</body>
</html>
```

> We will use the homepage `index.astro` file at the root to list out the blog articles.
> 
> For this we will use a `fetch` to call the Content Delivery API `content` endpoint and filter by contentType of 'article'.
> 
> The article objects will include the properties and content set in the CMS.
> 
> We can then loop through the articles and display a desired listing with a link to the article.
> 
> Replace the default contents of `index.astro` with the below:
> 
> ```astro title="src/pages/index.astro"
> ---
> import Layout from '../layouts/Layout.astro';
> const res = await fetch('http://localhost/umbraco/delivery/api/v2/content?filter=contentType:article');
> const articles = await res.json();
> ---
> <Layout>
> 	<h2>Blog Articles</h2>
> 	{
>         articles.items.map((article: any) => (
>             <div>
>                 <h3>{article.properties.title}</h3>
>                 <p>{article.properties.articleDate}</p>
>                 <a href={article.route.path}>Read more</a>
>             </div>
>         ))
>     }
> </Layout>
> ```

Reviewer: ``````suggestion
To create a list of blog posts, first `fetch` to call the Content Delivery API  `content` endpoint and filter by contentType of 'article'. The article objects will include the properties and content set in the CMS. You can then loop through the articles and display a each title with a link to its article.

Replace the default contents of `index.astro` with the following code:

```astro title="src/pages/index.astro"
---
import Layout from '../layouts/Layout.astro';
const res = await fetch('http://localhost/umbraco/delivery/api/v2/content?filter=contentType:article');
const articles = await res.json();
---
<Layout>
	<h2>Blog Articles</h2>
	{
        articles.items.map((article: any) => (
            <div>
                <h3>{article.properties.title}</h3>
                <p>{article.properties.articleDate}</p>
                <a href={article.route.path}>Read more</a>
            </div>
        ))
    }
</Layout>
```
``````

### Generating individual blog posts

> We will now create the file `[...slug].astro` file at the root of the `pages` directory.
> 
> The `[...slug]` convention here will be used to create the Dynamic Routes for our articles. Read more about Dynamic Routing [here](https://docs.astro.build/en/guides/routing/#dynamic-routes)
> 
> Here we use the `getStaticPaths()` function to return an an array of objects that represent out pages, in this case, our blog articles.
> 
> The `slug` property on `params` is used to generate the URL path of the page. This is defined by the `[...slug].astro` naming convention.
> 
> We are also passing a `props` object that will include the article returned from the API.
> 
> These properties can then be used within our markup by defining `const { article } = Astro.props;`
> 
> ```astro title="[...slug].astro"
> ---
> import Layout from '../layouts/Layout.astro';
> 
> export async function getStaticPaths() {
>     let data = await fetch("http://localhost/umbraco/delivery/api/v2/content?filter=contentType:article");
>     let articles = await data.json();
>     
>     return articles.items.map((article: any) => ({
>         params: { slug: article.route.path },
>         props: { article: article },
>     }));
> }
> 
> const { article } = Astro.props;
> ---
> 
> <Layout>
>   <h1>{article.properties.title}</h1>
>   <p>{article.properties.articleDate}</p>
>   <div set:html={article.properties.content?.markup}></div>
> </Layout>
> ```

Reviewer: ``````suggestion
Create the file `[...slug].astro` at the root of the `/pages/` directory. This file will be used to [generate the individual blog pages dynamically](https://docs.astro.build/en/guides/routing/#dynamic-routes).

Note that the `params` property, which generates the URL path of the page, contains `article.route.path` from the API fetch. Similarly, the `props` property must include the entire `article` itself so that you can access all the information in your CMS entry.

Add the following code to `[...slug].astro` which will create your individual blog post pages:

```astro title="[...slug].astro"
---
import Layout from '../layouts/Layout.astro';

export async function getStaticPaths() {
    let data = await fetch("http://localhost/umbraco/delivery/api/v2/content?filter=contentType:article");
    let articles = await data.json();
    
    return articles.items.map((article: any) => ({
        params: { slug: article.route.path },
        props: { article: article },
    }));
}

const { article } = Astro.props;
---

<Layout>
  <h1>{article.properties.title}</h1>
  <p>{article.properties.articleDate}</p>
  <div set:html={article.properties.content?.markup}></div>
</Layout>
```

Your project should now contain the following files:

<FileTree>
- src/
  - layouts/
    - Layout.astro
  - pages/
    - index.astro
    - **[...slug].astro**
</FileTree>
``````
I shortened the description of how `getStaticPaths()` works because that's all available at the link to our own dynamic routing content. What *is* maybe worth pointing out is how/what to include from the fetch call to the Content Delivery API so that the information is properly available!

> You should now be able to view your Umbraco-created content through the Astro front-end. Congratulations! 🚀

Reviewer: ```suggestion
With the dev server running, you should now be able to view your Umbraco-created content in your browser preview of your Astro project. Congratulations! 🚀
```

## Publishing your site

To deploy your site visit our [deployment guides](https://docs.astro.build/en/guides/deploy/) and follow the instructions for your preferred hosting provider.

## Local dev, HTTPS and self-signed SSL certificates

> When developing locally, be aware that Node (upon which Astro is built) won't accept self-signed certificates by default. This is a problem if using the Umbraco HTTPS endpoint locally, as any `fetch` queries will result in `fetch failed` with code `DEPTH_ZERO_SELF_SIGNED_CERT`.
> 
> To avoid this, either:
> 
> 1. Use the Umbraco HTTP endpoint for local dev
> 2. Set `NODE_TLS_REJECT_UNAUTHORIZED=0` within `astro.config.js`. An approach for this involves two files:
> 
> ```sh title=".env.development"
> NODE_TLS_REJECT_UNAUTHORIZED=0
> ```
> 
> ```astro title="astro.config.mjs"
> import { defineConfig } from 'astro/config';
> import { loadEnv } from "vite";
> 
> const { NODE_TLS_REJECT_UNAUTHORIZED } = loadEnv(process.env.NODE_ENV, process.cwd(), "");
> process.env.NODE_TLS_REJECT_UNAUTHORIZED = NODE_TLS_REJECT_UNAUTHORIZED;
> // https://astro.build/config
> export default defineConfig({});
> ```
> 
> This method is described in [this blog post](https://kjac.dev/posts/jamstack-for-free-with-azure-and-cloudflare/), with an [accompanying repo](https://github.com/kjac/UmbracoAzureCloudflare).

Reviewer: ``````suggestion
If you are using the Umbraco HTTPS endpoint locally, any `fetch` queries will result in `fetch failed` with code `DEPTH_ZERO_SELF_SIGNED_CERT`. This is because Node (upon which Astro is built) won't accept self-signed certificates by default. To avoid this, use the Umbraco HTTP endpoint for local dev.

Alternatively, you can set `NODE_TLS_REJECT_UNAUTHORIZED=0` in an `env.development` file and update `astro.config.js` as shown:

```sh title=".env.development"
NODE_TLS_REJECT_UNAUTHORIZED=0
```

```astro title="astro.config.mjs"
import { defineConfig } from 'astro/config';
import { loadEnv } from "vite";

const { NODE_TLS_REJECT_UNAUTHORIZED } = loadEnv(process.env.NODE_ENV, process.cwd(), "");
process.env.NODE_TLS_REJECT_UNAUTHORIZED = NODE_TLS_REJECT_UNAUTHORIZED;
// https://astro.build/config
export default defineConfig({});
```

This method is described in more detail in this [blog post showing how to configure your project for self-signed certificates](https://kjac.dev/posts/jamstack-for-free-with-azure-and-cloudflare/), with an [accompanying repo](https://github.com/kjac/UmbracoAzureCloudflare).
``````
I'm not sure if method one is *better* than method two, but it's sure shorter!  😄  So, given how unparallel the two methods are, I think it makes less sense to have an ordered list and more sense to say, 'Hey do this one great trick!" right in the description of the problem. THEN, describe the more involved process as an alternative for when the easy thing isn't possible or desirable.

---
title: Umbraco & Astro
description: Add content to your Astro project using Umbraco as a CMS
type: cms
stub: false
service: Umbraco
i18nReady: true
---
import { FileTree } from '@astrojs/starlight/components';

> import Grid from '~/components/FluidGrid.astro'
> import Card from '~/components/ShowcaseCard.astro'

Reviewer 2: ```suggestion
```

# Umbraco & Astro

[Umbraco CMS](https://umbraco.com/) is an open-source ASP.NET Core CMS. By default, Umbraco uses Razor pages for its front-end, but can be used as a headless CMS.

## Integrating with Astro

In this section you will use Umbraco's native [Content Delivery API](https://docs.umbraco.com/umbraco-cms/reference/content-delivery-api) to provide content to your Astro project.

## Prerequisites

To get started, you will need to have the following:

1. An Astro project - If you don’t have an Astro project yet, our [Installation guide](https://docs.astro.build/en/install/auto/) will get you up and running in no time.
2. An Umbraco (v12+) project - If you don’t have an Umbraco project yet, please see this [Installation guide](https://docs.umbraco.com/umbraco-cms/fundamentals/setup/install/).

## Setting up the Content Delivery API

To enable the Content Delivery API, update your Umbraco project's `appsettings.json` file:

> ```json

Reviewer 2: ``````suggestion
```json title="appsettings.json"
``````

    let articles = await data.json();
    
    return articles.items.map((article: any) => ({
        params: { slug: article.route.path },
        props: { article: article },
    }));
}

const { article } = Astro.props;
---

<Layout>
  <h1>{article.properties.title}</h1>
  <p>{article.properties.articleDate}</p>
  <div set:html={article.properties.content?.markup}></div>
</Layout>
```

You should now be able to view your Umbraco-created content through the Astro front-end. Congratulations! 🚀

## Publishing your site

To deploy your site visit our [deployment guides](https://docs.astro.build/en/guides/deploy/) and follow the instructions for your preferred hosting provider.

## Local dev, HTTPS and self-signed SSL certificates

When developing locally, be aware that Node (upon which Astro is built) won't accept self-signed certificates by default. This is a problem if using the Umbraco HTTPS endpoint locally, as any `fetch` queries will result in `fetch failed` with code `DEPTH_ZERO_SELF_SIGNED_CERT`.

To avoid this, either:

1. Use the Umbraco HTTP endpoint for local dev
2. Set `NODE_TLS_REJECT_UNAUTHORIZED=0` within `astro.config.js`. An approach for this involves two files:

> .env.development
> ```

Reviewer 2: ``````suggestion
```sh title=".env.development"
``````
